### PR TITLE
Enable keycloak SSO for podman login

### PR DESCRIPTION
**Pull request by newswangerd**
This adds a new authentication for basic authentication that attempts to log the user in using their keycloak username/password. If that fails, it falls back to regular django basic auth.

This new authentication class is configured as the default for basic auth when running keycloak, which enables keycloak basic auth on all the pulp apis, including the `/token/` api, which is required for `podman login`

Note, for this to work, Direct Access Grants must be enabled in keycloak.

Todo:
- [x] automate keycloak configuration
- [x] figure out how to validate jwt token

### CODE REVIEW

**Comment by newswangerd** [comment]
@chambridge how do I automate the keycloak setup to enable Direct Access Grants?

**Comment by chambridge** [comment]
> @chambridge how do I automate the keycloak setup to enable Direct Access Grants?

Pretty simple: https://docs.ansible.com/ansible/latest/collections/community/general/keycloak_client_module.html#parameter-direct_access_grants_enabled

Would just update here (https://github.com/ansible/galaxy_ng/blob/master/dev/standalone-keycloak/keycloak-playbook.yaml#L55) with `yes`

File: galaxy_ng/app/settings.py [hunk 1]
```diff
@@ -252,6 +252,12 @@
     SOCIAL_AUTH_KEYCLOAK_EXTRA_DATA = [('refresh_token', 'refresh_token'),
                                        (KEYCLOAK_ROLE_TOKEN_CLAIM, KEYCLOAK_ROLE_TOKEN_CLAIM), ]
 
+    # Enable basic auth using keycloak username/password
+    REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES = [
+        "rest_framework.authentication.SessionAuthentication",
+        "galaxy_ng.app.auth.keycloak.KeycloakBasicAuth",
+    ]
```
Comment (rochacbruno) [inline galaxy_ng/app/settings.py:259]: when/if #945 is merged, this settings is moved to dynaconf_hooks.py file https://github.com/ansible/galaxy_ng/pull/945/files#diff-51ac604a5e6e00d70270d768fd15bb83bca7f1d919f0471e0f43df8baba9d059R84

File: galaxy_ng/app/auth/keycloak.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            # should return a django user instance.
+            user = strategy.authenticate(backend, response=token_data)
+
+            if user is None:
+                raise exceptions.AuthenticationFailed(_("Authentication failed."))
+
+            return (user, None)
+
+        else:
+            # If keycloak basic auth fails, try regular basic auth.
```
Comment (ipanova) [inline galaxy_ng/app/auth/keycloak.py:55]: :+1:

File: galaxy_ng/app/settings.py (hunk 1 above)
Comment (rochacbruno) [reply galaxy_ng/app/settings.py:259]: @newswangerd move this to galaxy_ng/app/dynaconf_hooks.py https://github.com/ansible/galaxy_ng/blob/master/galaxy_ng/app/dynaconf_hooks.py#L84

there you add it to a dictionary
```py
data["REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES"] = [
    "rest_framework.authentication.SessionAuthentication",
    "galaxy_ng.app.auth.keycloak.KeycloakBasicAuth",
]
```

**Review by rochacbruno — CHANGES_REQUESTED** [review]

**Review by rochacbruno — APPROVED** [review]

**Pull request merged**